crates/wasm-encoder/src/imports.rs

```diff
@@ -46,36 +53,38 @@ impl ImportSection {
         self.num_added == 0
     }
 
-    /// Define an import.
+    /// Define an import in the import section.
+    ///
+    /// When encoding modules, the import's module name is required.
+    /// When encoding adapter modules and components, the import's module name must be `None`.
     pub fn import(
         &mut self,
-        module: &str,
-        name: Option<&str>,
-        ty: impl Into<EntityType>,
+        module: Option<&str>,
+        name: &str,
+        ty: impl Into<TypeRef>,
     ) -> &mut Self {
-        self.bytes.extend(encoders::str(module));
-        match name {
-            Some(name) => self.bytes.extend(encoders::str(name)),
-            None => {
-                self.bytes.push(0x00);
-                self.bytes.push(0xff);
+        match self.format {
+            SectionEncodingFormat::Module => {
+                // The module name is required
+                let module = module.expect("module name is required for module encoding");
+                self.bytes.extend(encoders::str(module));
+                self.bytes.extend(encoders::str(name));
+            }
+            SectionEncodingFormat::AdapterModule | SectionEncodingFormat::Component => {
+                assert!(
+                    module.is_none(),
+                    "module name cannot be provided for adapter module or component encoding"
+                );
+                self.bytes.extend(encoders::str(name));
             }
         }
-        ty.into().encode(&mut self.bytes);
+        ty.into().encode(self.format, &mut self.bytes);
         self.num_added += 1;
         self
     }
```

Comment:
For the import section I would imagine that because the encoding is different between core wasm and the component model that this would be two methods, one for each encoding choice, instead of one method with various optional parameters.

Reply:
Agreed, this can be two different methods. I'll fix.

Reply:
I've added an `import_by_name` function specifically for encoding functions (totally open to a better name than that one).